[HtmlSanitizer] Add functions to handle operations on multiple attributes and/or elements

EdouardCourty opened the pull request:
Add functions to handle operations on multiple attributes or elements at the same time

| Q             | A
| ------------- | ---
| Branch?       | 7.1
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | None
| License       | MIT

This branch brings supports for array parameters to handle mass allow / block / drop of elements and/or attributes.

Here is a code screenshot that shows how these news methods work:

```php
$config = new HtmlSanitizerConfig();

$attributes = ['height', 'width', 'src'];
$elements = ['img',' iframe'];

/* Current way */
        
foreach ($elements as $element) {
    $config = $config->allowElement($element, $attributes);
}

/* New way */

$config = $config->allowElements($elements, $attributes);
```

EdouardCourty committed:
[HtmlSanitizer] Add functions to handle operations on multiple attributes or elements at the same time

EdouardCourty commented:
By the way, why is the `HtmlSanitizerConfig` class using `clone` statements in all its methods?
![image](https://github.com/symfony/symfony/assets/37371516/0c74f9dd-1b7a-404f-858b-828a78bf49ba)

stof commented:
@EdouardCourty being the Config object is immutable (which means that even if you keep a reference to the Config object, you cannot mutate the configuration of a Sanitizer that has been instantiated with it)

stof commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerConfig.php:
```diff
@@ -274,6 +274,29 @@ public function allowElement(string $element, array|string $allowedAttributes =
+     *
+     * Allowed elements are elements the sanitizer should retain from the input.
+     *
+     * A list of allowed attributes for this element can be passed as a second argument.
```
`this element` does not make sense in this sentence, as you pass multiple ones.

stof commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerConfig.php:
```diff
@@ -292,6 +315,23 @@ public function blockElement(string $element): static
+        $clone = clone $this;
+
+        foreach ($elements as $element) {
+            $clone = $clone->blockElement($element);
```
Instead of using the public API (which will clone the config many times), you should modify the state of the clone directly, which would allow cloning only once. Otherwise, there is no real benefit for those methods

EdouardCourty replied:
Hi, I wrapped the logic inside private functions that I can reuse in the loops and in the base functions

stof commented on src/Symfony/Component/HtmlSanitizer/HtmlSanitizerConfig.php:
```diff
@@ -367,6 +454,32 @@ public function dropAttribute(string $attribute, array|string $droppedElements):
+     * automatically. This method let you drop attributes that were allowed earlier
+     * in the configuration.
+     *
+     * @param list<string>        $attributes
```
should be `string[]`. As the method iterate over the attributes without caring about keys at all, there is no reason to force the caller to pass a list.

EdouardCourty replied:
Hi, I agree!
The rest of the methods are documented this way, which is why I thought I'd stay in the same style.

EdouardCourty commented:
> @EdouardCourty being the Config object is immutable (which means that even if you keep a reference to the Config object, you cannot mutate the configuration of a Sanitizer that has been instantiated with it)

Thanks for your reply, I'll fix the PR asap

EdouardCourty committed:
[HtmlSanitizer] Add functions to handle operations on multiple attributes or elements at the same time
- Wrapped some logic in private functions to stay DRY
- Fixed PHPDoc comments
- Fixed typo
